File: mesh3D/Fortran3D/examples/example_polynomial.f90
PROGRAM example_polynomial

USE TypesDef, ONLY: dp
USE OptimalPolyCuba3D_Module, ONLY: OptimalPolyCuba3D

IMPLICIT NONE

!***********************************************************************
!
!   Esempio:
!       Cubatura su dominio cubico
!
!   Descrizione:
!       Questo esempio dimostra l'utilizzo del metodo OptimalPolyCuba3D
!       per l'integrazione di funzioni polinomiali su un cubo, rappresentato
!       mediante una mesh superficiale triangolare chiusa e orientata
!       secondo le normali esterne.
!
!       Vengono considerate tre funzioni integrande di diverso grado
!       polinomiale e i risultati numerici vengono confrontati con
!       i corrispondenti valori analitici.
!
!***********************************************************************


!***********************************************************************
!   Dichiarazione delle variabili
!***********************************************************************

INTEGER                         :: ade
INTEGER                         :: n_vertici
INTEGER                         :: n_facce
INTEGER                         :: i

REAL(dp)                        :: Integrale
REAL(dp)                        :: I_exact
REAL(dp)                        :: error_abs
REAL(dp)                        :: elapsedTime
REAL(dp)                        :: t_start
REAL(dp)                        :: t_end

REAL(dp), ALLOCATABLE           :: vertices(:,:)
INTEGER, ALLOCATABLE            :: facets(:,:)

REAL(dp), ALLOCATABLE           :: XYZ(:,:)
REAL(dp), ALLOCATABLE           :: W(:)
REAL(dp), ALLOCATABLE           :: fXYZ(:)


!***********************************************************************
!   Parametri e caricamento mesh
!***********************************************************************

WRITE(*,'(A)')
WRITE(*,'(A)') '**************************************************************'
WRITE(*,'(A)') '                 OPTIMALPOLYCUBA3D'
WRITE(*,'(A)') '                 Cubatura sul Cubo'
WRITE(*,'(A)') '**************************************************************'
WRITE(*,'(A)')


!***********************************************************************
!   Vertici e facce del cubo [-1,1]^3
!***********************************************************************

ALLOCATE(vertices(8,3))

vertices = RESHAPE([ &
    -1.0_dp, -1.0_dp, -1.0_dp, &
     1.0_dp, -1.0_dp, -1.0_dp, &
     1.0_dp,  1.0_dp, -1.0_dp, &
    -1.0_dp,  1.0_dp, -1.0_dp, &
    -1.0_dp, -1.0_dp,  1.0_dp, &
     1.0_dp, -1.0_dp,  1.0_dp, &
     1.0_dp,  1.0_dp,  1.0_dp, &
    -1.0_dp,  1.0_dp,  1.0_dp  &
    ], [8,3])

ALLOCATE(facets(12,3))

facets = RESHAPE([ &
    1, 3, 2, &
    1, 4, 3, &
    5, 6, 7, &
    5, 7, 8, &
    1, 2, 6, &
    1, 6, 5, &
    4, 8, 7, &
    4, 7, 3, &
    1, 5, 8, &
    1, 8, 4, &
    2, 3, 7, &
    2, 7, 6  &
    ], [12,3])


n_vertici = SIZE(vertices,1)
n_facce   = SIZE(facets,1)

WRITE(*,'(A,I0)') 'Numero di vertici:     ', n_vertici
WRITE(*,'(A,I0)') 'Numero di facce:       ', n_facce

!***********************************************************************
!   Definizione funzione integranda f1
!***********************************************************************

ade = 1
I_exact = 8.0_dp

WRITE(*,'(A)')
WRITE(*,'(A)') '--------------------------------------------------------------'
WRITE(*,'(A)') 'Funzione integranda 1'
WRITE(*,'(A)') '--------------------------------------------------------------'
WRITE(*,'(A,I0)')    'Ade:                   ', ade
WRITE(*,'(A)')       'Funzione integranda:   f_1(x,y,z) = 1'
WRITE(*,'(A,ES22.15)') 'Integrale esatto:      ', I_exact

!***********************************************************************
!   Inizio regola di cubatura
!***********************************************************************

WRITE(*,'(A)')
WRITE(*,'(A)') 'Inizio Cubatura...'

CALL CPU_TIME(t_start)
CALL OptimalPolyCuba3D(ade, vertices, facets, XYZ, W)
CALL CPU_TIME(t_end)

elapsedTime = t_end - t_start

!***********************************************************************
!   Valutazione della funzione integranda
!***********************************************************************

ALLOCATE(fXYZ(SIZE(W)))

DO i = 1, SIZE(W)
    fXYZ(i) = f1(XYZ(i,1), XYZ(i,2), XYZ(i,3))
END DO

Integrale = DOT_PRODUCT(W, fXYZ)
DEALLOCATE(fXYZ)

!***********************************************************************
!   Visualizzazione risultati
!***********************************************************************

WRITE(*,'(A)') 'Fine Cubatura...'

error_abs = ABS(Integrale - I_exact)

WRITE(*,'(A)')
WRITE(*,'(A,ES22.15)') 'Integrale numerico:    ', Integrale
WRITE(*,'(A,ES14.6)')  'Errore assoluto:       ', error_abs
WRITE(*,'(A,ES14.6,A)') 'Tempo di calcolo:      ', elapsedTime, ' s'


!***********************************************************************
!   Definizione funzione integranda f2
!***********************************************************************

ade = 2
I_exact = 8.0_dp

WRITE(*,'(A)')
WRITE(*,'(A)') '--------------------------------------------------------------'
WRITE(*,'(A)') 'Funzione integranda 2'
WRITE(*,'(A)') '--------------------------------------------------------------'
WRITE(*,'(A,I0)')    'Ade:                   ', ade
WRITE(*,'(A)')       'Funzione integranda:   f_2(x,y,z) = x^2 + y^2 + z^2'
WRITE(*,'(A,ES22.15)') 'Integrale esatto:      ', I_exact

!***********************************************************************
!   Inizio regola di cubatura
!***********************************************************************

WRITE(*,'(A)')
WRITE(*,'(A)') 'Inizio Cubatura...'

CALL CPU_TIME(t_start)
CALL OptimalPolyCuba3D(ade, vertices, facets, XYZ, W)
CALL CPU_TIME(t_end)

elapsedTime = t_end - t_start

!***********************************************************************
!   Valutazione della funzione integranda
!***********************************************************************

IF (ALLOCATED(fXYZ)) DEALLOCATE(fXYZ)
ALLOCATE(fXYZ(SIZE(W)))

DO i = 1, SIZE(W)
    fXYZ(i) = f2(XYZ(i,1), XYZ(i,2), XYZ(i,3))
END DO

Integrale = DOT_PRODUCT(W, fXYZ)

DEALLOCATE(fXYZ)

!***********************************************************************
!   Visualizzazione risultati
!***********************************************************************

WRITE(*,'(A)') 'Fine Cubatura...'

error_abs = ABS(Integrale - I_exact)

WRITE(*,'(A)')
WRITE(*,'(A,ES22.15)') 'Integrale numerico:    ', Integrale
WRITE(*,'(A,ES14.6)')  'Errore assoluto:       ', error_abs
WRITE(*,'(A,ES14.6,A)') 'Tempo di calcolo:      ', elapsedTime, ' s'

!***********************************************************************
!   Definizione funzione integranda f3
!***********************************************************************

ade = 6

I_exact = 8.0_dp / 27.0_dp

WRITE(*,'(A)')
WRITE(*,'(A)') '--------------------------------------------------------------'
WRITE(*,'(A)') 'Funzione integranda 3'
WRITE(*,'(A)') '--------------------------------------------------------------'
WRITE(*,'(A,I0)')    'Ade:                   ', ade
WRITE(*,'(A)')       'Funzione integranda:   f_3(x,y,z) = x^2*y^2*z^2'
WRITE(*,'(A,ES22.15)') 'Integrale esatto:      ', I_exact


!***********************************************************************
!   Inizio regola di cubatura
!***********************************************************************

WRITE(*,'(A)')
WRITE(*,'(A)') 'Inizio Cubatura...'

CALL CPU_TIME(t_start)
CALL OptimalPolyCuba3D(ade, vertices, facets, XYZ, W)
CALL CPU_TIME(t_end)

elapsedTime = t_end - t_start

!***********************************************************************
!   Valutazione della funzione integranda
!***********************************************************************

IF (ALLOCATED(fXYZ)) DEALLOCATE(fXYZ)
ALLOCATE(fXYZ(SIZE(W)))

DO i = 1, SIZE(W)
    fXYZ(i) = f3(XYZ(i,1), XYZ(i,2), XYZ(i,3))
END DO

Integrale = DOT_PRODUCT(W, fXYZ)

DEALLOCATE(fXYZ)

!***********************************************************************
!   Visualizzazione risultati
!***********************************************************************

WRITE(*,'(A)') 'Fine Cubatura...'

error_abs = ABS(Integrale - I_exact)

WRITE(*,'(A)')
WRITE(*,'(A,ES22.15)') 'Integrale numerico:    ', Integrale
WRITE(*,'(A,ES14.6)')  'Errore assoluto:       ', error_abs
WRITE(*,'(A,ES14.6,A)') 'Tempo di calcolo:      ', elapsedTime, ' s'

!***********************************************************************
!   Deallocazione memoria
!***********************************************************************

IF (ALLOCATED(vertices)) DEALLOCATE(vertices)
IF (ALLOCATED(facets))   DEALLOCATE(facets)
IF (ALLOCATED(XYZ))      DEALLOCATE(XYZ)
IF (ALLOCATED(W))        DEALLOCATE(W)

CONTAINS

!***********************************************************************
!   Funzione integranda f1
!***********************************************************************

REAL(dp) FUNCTION f1(x,y,z)

    IMPLICIT NONE

    REAL(dp), INTENT(IN) :: x
    REAL(dp), INTENT(IN) :: y
    REAL(dp), INTENT(IN) :: z

    f1 = 1.0_dp

END FUNCTION f1


!***********************************************************************
!   Funzione integranda f2
!***********************************************************************

REAL(dp) FUNCTION f2(x,y,z)

    IMPLICIT NONE

    REAL(dp), INTENT(IN) :: x
    REAL(dp), INTENT(IN) :: y
    REAL(dp), INTENT(IN) :: z

    f2 = x**2 + y**2 + z**2

END FUNCTION f2


!***********************************************************************
!   Funzione integranda f3
!***********************************************************************

REAL(dp) FUNCTION f3(x,y,z)

    IMPLICIT NONE

    REAL(dp), INTENT(IN) :: x
    REAL(dp), INTENT(IN) :: y
    REAL(dp), INTENT(IN) :: z

    f3 = x**2 * y**2 * z**2

END FUNCTION f3


END PROGRAM example_polynomial

File: Mesh3D/Matlab3D/examples/example_polynomial.m
clear;
clc;
close all;

projectRoot = fileparts(fileparts(mfilename('fullpath')));
addpath(projectRoot);
addpath(fullfile(projectRoot, 'src'));

%**************************************************************************
%
%   Esempio:
%       Cubatura su dominio cubico
%
%   Descrizione:
%       Questo esempio dimostra l'utilizzo del metodo OptimalPolyCuba3D
%       per l'integrazione di funzioni polinomiali su un cubo, rappresentato
%       mediante una mesh superficiale triangolare chiusa e orientata
%       secondo le normali esterne.
%
%       Vengono considerate tre funzioni integrande di diverso grado
%       polinomiale e i risultati numerici vengono confrontati con
%       i corrispondenti valori analitici.
%
%**************************************************************************

%**************************************************************************
%  Parametri e caricamento mesh
%**************************************************************************

fprintf('\n');
fprintf('**************************************************************\n');
fprintf('                 OPTIMALPOLYCUBA3D\n');
fprintf('                 Cubatura sul Cubo\n');
fprintf('**************************************************************\n');
fprintf('\n');

% Vertici del cubo [-1,1]^3
vertices = [
    -1 -1 -1;
     1 -1 -1;
     1  1 -1;
    -1  1 -1;
    -1 -1  1;
     1 -1  1;
     1  1  1;
    -1  1  1
];

% Facce triangolari della mesh superficiale.
% I vertici di ogni triangolo sono ordinati in senso antiorario
% se osservati dall'esterno del dominio, garantendo normali esterne.

facets = [
    1 3 2;
    1 4 3;

    5 6 7;
    5 7 8;

    1 2 6;
    1 6 5;

    4 8 7;
    4 7 3;

    1 5 8;
    1 8 4;

    2 3 7;
    2 7 6
];

fprintf('Numero di vertici:     %d\n', size(vertices,1));
fprintf('Numero di facce:       %d\n', size(facets,1));

%**************************************************************************
% Plot dominio poliedrale
%**************************************************************************

figure('Color','w');

patch('Vertices', vertices, ...
      'Faces', facets, ...
      'FaceColor', [0.7 0.7 0.7], ...
      'EdgeColor', [0.2 0.2 0.2], ...
      'FaceAlpha', 0.65);

axis equal;
grid on;
box on;

xlabel('$x$', 'Interpreter','latex');
ylabel('$y$', 'Interpreter','latex');
zlabel('$z$', 'Interpreter','latex');

title('Dominio Poliedrale', 'Interpreter','latex');

view(3);

%**************************************************************************
% Definizione funzione integranda f1
%**************************************************************************

ade = 1;

f1 = @(x,y,z) ones(size(x));
f1_string = 'f_1(x,y,z) = 1';

I_exact = 8;

fprintf('\n');
fprintf('--------------------------------------------------------------\n');
fprintf('Funzione integranda 1\n');
fprintf('--------------------------------------------------------------\n');
fprintf('Ade:                   %d\n', ade);
fprintf('Funzione integranda:   %s\n', f1_string);
fprintf('Integrale esatto:      %.15e\n', I_exact);

%**************************************************************************
% Inizio regola di cubatura
%**************************************************************************

fprintf('\n');
fprintf('Inizio Cubatura...\n');

tic;
[XYZ, W] = OPC3D(ade, vertices, facets, "D");
%[XYZ, W] = OPC3D(ade, vertices, facets, "GJ");
elapsedTime = toc;

I = W' * f1(XYZ(:,1), XYZ(:,2), XYZ(:,3));

fprintf('Fine Cubatura...\n');

error_abs = abs(I - I_exact);

fprintf('\n');
fprintf('Integrale numerico:    %.15e\n', I);
fprintf('Errore assoluto:       %.6e\n', error_abs);
fprintf('Tempo di calcolo:      %.6e s\n', elapsedTime);

%**************************************************************************
% Definizione funzione integranda f2
%**************************************************************************

ade = 2;

f2 = @(x,y,z) x.^2 + y.^2 + z.^2;
f2_string = 'f_2(x,y,z) = x^2 + y^2 + z^2';

I_exact = 8;

fprintf('\n');
fprintf('--------------------------------------------------------------\n');
fprintf('Funzione integranda 2\n');
fprintf('--------------------------------------------------------------\n');
fprintf('Ade:                   %d\n', ade);
fprintf('Funzione integranda:   %s\n', f2_string);
fprintf('Integrale esatto:      %.15e\n', I_exact);

%**************************************************************************
% Inizio regola di cubatura
%**************************************************************************

fprintf('\n');
fprintf('Inizio Cubatura...\n');

tic;
[XYZ, W] = OPC3D(ade, vertices, facets, "D");
%[XYZ, W] = OPC3D(ade, vertices, facets, "GJ");
elapsedTime = toc;


I = W' * f2(XYZ(:,1), XYZ(:,2), XYZ(:,3));

fprintf('Fine Cubatura...\n');

error_abs = abs(I - I_exact);

fprintf('\n');
fprintf('Integrale numerico:    %.15e\n', I);
fprintf('Errore assoluto:       %.6e\n', error_abs);
fprintf('Tempo di calcolo:      %.6e s\n', elapsedTime);

%**************************************************************************
% Definizione funzione integranda f3
%**************************************************************************

ade = 6;

f3 = @(x,y,z) x.^2 .* y.^2 .* z.^2;
f3_string = 'f_3(x,y,z) = x^2*y^2*z^2';

I_exact = 8/27;

fprintf('\n');
fprintf('--------------------------------------------------------------\n');
fprintf('Funzione integranda 3\n');
fprintf('--------------------------------------------------------------\n');
fprintf('Ade:                   %d\n', ade);
fprintf('Funzione integranda:   %s\n', f3_string);
fprintf('Integrale esatto:      %.15e\n', I_exact);

%**************************************************************************
% Inizio regola di cubatura
%**************************************************************************

fprintf('\n');
fprintf('Inizio Cubatura...\n');

tic;
[XYZ, W] = OPC3D(ade, vertices, facets, "D");
%[XYZ, W] = OPC3D(ade, vertices, facets, "GJ");
elapsedTime = toc;

I = W' * f3(XYZ(:,1), XYZ(:,2), XYZ(:,3));

fprintf('Fine Cubatura...\n');

error_abs = abs(I - I_exact);

fprintf('\n');
fprintf('Integrale numerico:    %.15e\n', I);
fprintf('Errore assoluto:       %.6e\n', error_abs);
fprintf('Tempo di calcolo:      %.6e s\n', elapsedTime);